### PR TITLE
Cleanup of model system, taking advantage of java8.

### PR DESCRIPTION
**Pull request by RainWarrior**
Also fixed some null annotation issues.
I think I didn't break binary compatibility (much), but would be nice if somebody else looked at it too.
Also, replacing guava's Optional/Function with java8 versions should probably happen at some point, but that would obviously be a breaking change, so it's not included here.

### CODE REVIEW

**Comment by loordgek** [comment]
replacing guava's Optional/Function with java8
guava's extend java now 
https://github.com/google/guava/wiki/Release21
> Function, Predicate and Supplier: changed to extend the new java.util.function interfaces with the same names.

**Comment by kashike** [comment]
@loordgek  You can pass guava predicates wherever java ones are accepted, not the other way around.

File: src/main/java/net/minecraftforge/client/model/ModelLoader.java [hunk 1]
```diff
@@ -310,20 +288,8 @@ protected void loadItemModels()
...
-            }
-        }));
-        Collections.sort(items, new Comparator<Item>()
-        {
-            public int compare(Item i1, Item i2)
-            {
-                return i1.getRegistryName().toString().compareTo(i2.getRegistryName().toString());
-            }
-        });
+        List<Item> items = Lists.newArrayList(Iterables.filter(Item.REGISTRY, item -> item.getRegistryName() != null));
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/model/ModelLoader.java:291]: You should use method reference of `Objects#nonNull` to make it neater.

File: src/main/java/net/minecraftforge/client/model/ModelLoader.java [hunk 2]
```diff
@@ -310,20 +288,8 @@ protected void loadItemModels()
...
+        Collections.sort(items, (i1, i2) -> i1.getRegistryName().toString().compareTo(i2.getRegistryName().toString()));
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/model/ModelLoader.java:292]: You can use `Comparator#comparing(Function)` to compare the registry names.

File: src/main/java/net/minecraftforge/client/model/ModelLoader.java (hunk 1 above)
Comment (liach) [reply src/main/java/net/minecraftforge/client/model/ModelLoader.java:291]: This comment is invalid, please ignore.

**Comment by covers1624** [comment]
Looks really good to me. Honestly, Since forge hasn't finalized for 1.12 yet it should be fine to do all the breaking changes we need right?

**Comment by williewillus** [comment]
Yeah now's the time if any

**Comment by RainWarrior** [comment]
https://github.com/MinecraftForge/MinecraftForge/commit/dc043ac79f90a9b0da5efff64bb02a4a7301c242

**Pull request closed without merging**